lnwallet/chanfunding/psbt_assembler.go

```diff
@@ -0,0 +1,372 @@
+			err)
+	}
+
+	return addr.EncodeAddress(), out.Value, buf.Bytes(), nil
+}
+
+// Verify makes sure the PSBT that is given to the intent has an output that
+// sends to the channel funding multisig address with the correct amount. A
+// simple check that there is a large enough input present is also performed.
+func (i *PsbtIntent) Verify(rawPsbt []byte) error {
+	// We want the PSBT in the raw binary format, base64 is only used when
+	// a direct user interaction is involved.
+	packet, err := psbt.NewFromRawBytes(bytes.NewReader(rawPsbt), false)
+	if err != nil {
+		return fmt.Errorf("error parseing PSBT: %v", err)
+	}
+
+	// Try to locate the channel funding multisig output.
+	_, expectedOutput, err := i.FundingOutput()
+	if err != nil {
+		return fmt.Errorf("funding output cannot be created: %v", err)
+	}
+	outputFound := false
+	for _, out := range packet.UnsignedTx.TxOut {
+		if reflect.DeepEqual(out, expectedOutput) {
+			outputFound = true
+		}
+	}
+	if !outputFound {
+		return fmt.Errorf("funding output not found in PSBT")
+	}
+
+	// At least one input needs to be specified. We don't want to dive
+	// deeper into input validation for the moment as that involves many
+	// intricacies. We see it as the input adder's job to make sure the
+	// amount is sufficient and that a change output is added.
+	if len(packet.UnsignedTx.TxIn) == 0 {
+		return fmt.Errorf("PSBT has no inputs")
+	}
+
+	i.PendingPsbt = packet
+	i.State = PsbtVerified
+	return nil
+}
+
+// Finalize makes sure the final PSBT that is given to the intent is fully valid
+// and signed but still contains the same UTXOs and outputs as the pending
+// transaction we previously verified. If everything checks out, the funding
+// manager is informed that the channel can now be opened and the funding
+// transaction be broadcast.
+func (i *PsbtIntent) Finalize(rawPsbt []byte) error {
+	// Parse the raw binary format and make sure the PSBT itself thinks it's
+	// finalized and ready to be broadcast.
+	packet, err := psbt.NewFromRawBytes(bytes.NewReader(rawPsbt), false)
+	if err != nil {
+		return fmt.Errorf("error parseing PSBT: %v", err)
+	}
+	err = psbt.MaybeFinalizeAll(packet)
+	if err != nil {
+		return fmt.Errorf("error finalizing PSBT: %v", err)
+	}
+	if !packet.IsComplete() {
+		return fmt.Errorf("PSBT not complete")
+	}
+	_, err = psbt.Extract(packet)
+	if err != nil {
+		return fmt.Errorf("unable to extract funding TX: %v", err)
+	}
+
+	// Do a rudimentary check that this is still the same PSBT that we
+	// verified in the previous step. This is to protect the user from
+	// unwanted modifications.
+	if i.PendingPsbt == nil {
+		return fmt.Errorf("PSBT was not verified first")
+	}
+	verifiedOuts := i.PendingPsbt.UnsignedTx.TxOut
+	if len(packet.UnsignedTx.TxOut) != len(verifiedOuts) {
+		return fmt.Errorf("number of outputs differ from verified PSBT")
+	}
+	for idx, out := range packet.UnsignedTx.TxOut {
+		if !reflect.DeepEqual(out, verifiedOuts[idx]) {
```

Comment:
Similar thing here about output value (can likely be coupled into a helper function). Also, if we want to go even further here, we could do a full equality check (removed witness/sigScript) between this an the prior one we extracted.